From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: celenity <celenity@celenity.dev>
Date: Sun, 05 Jul 2026 20:19:20 +0000
Subject: [PATCH] Gecko - Remote Settings Blocker

Prevents Remote Settings from downloading unwanted collections.

Adapted from LibreWolf: https://codeberg.org/librewolf/source/src/branch/main/patches/rs-blocker.patch

Signed-off-by: celenity <celenity@celenity.dev>
---
diff --git a/services/settings/RemoteSettings.worker.mjs b/services/settings/RemoteSettings.worker.mjs
index 3e4e915c13ed..1df034046bd0 100644
--- a/services/settings/RemoteSettings.worker.mjs
+++ b/services/settings/RemoteSettings.worker.mjs
@@ -64,10 +64,12 @@ const Agent = {
    * @param {string} collection
    * @returns {int} Number of records loaded from dump or -1 if no dump found.
    */
-  async importJSONDump(bucket, collection) {
+  async importJSONDump(bucket, collection, allowedCollections, allowedCollectionsFromDump) {
     const { data: records, timestamp } = await SharedUtils.loadJSONDump(
       bucket,
-      collection
+      collection,
+      allowedCollections,
+      allowedCollectionsFromDump
     );
     if (records === null) {
       // Return -1 if file is missing.
diff --git a/services/settings/RemoteSettingsClient.sys.mjs b/services/settings/RemoteSettingsClient.sys.mjs
index f84cdf9796925..4b4424845434b 100644
--- a/services/settings/RemoteSettingsClient.sys.mjs
+++ b/services/settings/RemoteSettingsClient.sys.mjs
@@ -209,12 +209,21 @@ class AttachmentDownloader extends Downloader {
     return cacheImpl;
   }
 
+  get identifier() {
+    return this._client.identifier;
+  }
+
   /**
    * Download attachment and report Telemetry on failure.
    *
    * @see Downloader.download
    */
   async download(record, options) {
+    if (!lazy.SharedUtils.isCollectionAllowed(this.bucketName, this.collectionName, lazy.Utils.allowedCollections)) {
+      throw Error(
+        `Download attempt to RS collection "${this.identifier}" was blocked.`
+      );
+    }
     await lazy.UptakeTelemetry.report(
       lazy.UptakeTelemetry.STATUS.DOWNLOAD_START,
       {
@@ -586,7 +595,9 @@ export class RemoteSettingsClient extends EventEmitter {
       }
       ({ data } = await lazy.SharedUtils.loadJSONDump(
         this.bucketName,
-        this.collectionName
+        this.collectionName,
+        lazy.Utils.allowedCollections,
+        lazy.Utils.allowedCollectionsFromDump
       ));
       if (data !== null) {
         lazy.console.info(`${this.identifier} falling back to JSON dump`);
@@ -656,6 +667,10 @@ export class RemoteSettingsClient extends EventEmitter {
       return;
     }
 
+    if (!lazy.SharedUtils.isCollectionAllowed(this.bucketName, this.collectionName, lazy.Utils.allowedCollections)) {
+      return;
+    }
+
     // We want to know which timestamp we are expected to obtain in order to leverage
     // cache busting. We don't provide ETag because we don't want a 304.
     const { changes } = await lazy.Utils.fetchLatestChanges(
@@ -1040,7 +1055,9 @@ export class RemoteSettingsClient extends EventEmitter {
     lazy.console.info(`${this.identifier} try to restore dump`);
     const result = await lazy.RemoteSettingsWorker.importJSONDump(
       this.bucketName,
-      this.collectionName
+      this.collectionName,
+      lazy.Utils.allowedCollections,
+      lazy.Utils.allowedCollectionsFromDump
     );
     if (result < 0) {
       lazy.console.debug(`${this.identifier} no dump available`);
diff --git a/services/settings/RemoteSettingsWorker.sys.mjs b/services/settings/RemoteSettingsWorker.sys.mjs
index 57bfb3f3d544..874b35abff18 100644
--- a/services/settings/RemoteSettingsWorker.sys.mjs
+++ b/services/settings/RemoteSettingsWorker.sys.mjs
@@ -162,8 +162,8 @@ class Worker {
     ]);
   }
 
-  async importJSONDump(bucket, collection) {
-    return this._execute("importJSONDump", [bucket, collection], {
+  async importJSONDump(bucket, collection, allowedCollections, allowedCollectionsFromDump) {
+    return this._execute("importJSONDump", [bucket, collection, allowedCollections, allowedCollectionsFromDump], {
       mustComplete: true,
     });
   }
diff --git a/services/settings/SharedUtils.sys.mjs b/services/settings/SharedUtils.sys.mjs
index e5b20d9b9dd6e..1f97726c31999 100644
--- a/services/settings/SharedUtils.sys.mjs
+++ b/services/settings/SharedUtils.sys.mjs
@@ -35,7 +35,10 @@ export var SharedUtils = {
    * @param {string}  bucket
    * @param {string}  collection
    */
-  async loadJSONDump(bucket, collection) {
+  async loadJSONDump(bucket, collection, allowedCollections, allowedCollectionsFromDump) {
+    if (!this.isCollectionAllowedFromDump(bucket, collection, allowedCollections, allowedCollectionsFromDump)) {
+      return { data: null, timestamp: null };
+    }
     // When using the preview bucket, we still want to load the main dump.
     // But we store it locally in the preview bucket.
     const jsonBucket = bucket.replace("-preview", "");
@@ -50,4 +53,37 @@ export var SharedUtils = {
     // Will throw if JSON is invalid.
     return response.json();
   },
+
+  _isCollectionAllowedInternal(bucket, collection, allowedCollections) {
+    return (
+      allowedCollections.includes(`${bucket}/${collection}`) ||
+      allowedCollections.includes(`${bucket}/*`) ||
+      allowedCollections.includes("*")
+    );
+  },
+
+  isCollectionAllowed(bucket, collection, allowedCollections) {
+    if (this._isCollectionAllowedInternal(bucket, collection, allowedCollections)) {
+      return true;
+    }
+    // eslint-disable-next-line no-console
+    console.debug(
+      `Connection attempt to RS collection "${bucket}/${collection}" was blocked/filtered.`
+    );
+    return false;
+  },
+
+  isCollectionAllowedFromDump(bucket, collection, allowedCollections, allowedCollectionsFromDump) {
+    if (
+      this._isCollectionAllowedInternal(bucket, collection, allowedCollectionsFromDump) ||
+      this._isCollectionAllowedInternal(bucket, collection, allowedCollections)
+    ) {
+      return true;
+    }
+    // eslint-disable-next-line no-console
+    console.debug(
+      `Access attempt to RS collection "${bucket}/${collection}" from local dump was blocked/filtered.`
+    );
+    return false;
+  },
 };
diff --git a/services/settings/Utils.sys.mjs b/services/settings/Utils.sys.mjs
index 70f77c9fcbbb6..58d1e2f8d8e50 100644
--- a/services/settings/Utils.sys.mjs
+++ b/services/settings/Utils.sys.mjs
@@ -11,6 +11,7 @@ const lazy = {};
 
 ChromeUtils.defineESModuleGetters(lazy, {
   SharedUtils: "resource://services-settings/SharedUtils.sys.mjs",
+  IFPrefUtils: "moz-src:///ironfox/utils/IFPrefUtils.sys.mjs",
 });
 
 XPCOMUtils.defineLazyServiceGetter(
@@ -99,6 +100,18 @@ XPCOMUtils.defineLazyPreferenceGetter(
   ""
 );
 
+ChromeUtils.defineLazyGetter(lazy, "allowedCollections", () =>
+  lazy.IFPrefUtils
+    .getStringPref("browser.ironfox.services.settings.allowedCollections")
+    .split(",")
+);
+
+ChromeUtils.defineLazyGetter(lazy, "allowedCollectionsFromDump", () =>
+  lazy.IFPrefUtils
+    .getStringPref("browser.ironfox.services.settings.allowedCollectionsFromDump")
+    .split(",")
+);
+
 function _isUndefined(value) {
   return typeof value === "undefined";
 }
@@ -111,6 +124,14 @@ export var Utils = {
       : AppConstants.REMOTE_SETTINGS_SERVER_URLS[0];
   },
 
+  get allowedCollections() {
+    return lazy.allowedCollections;
+  },
+
+  get allowedCollectionsFromDump() {
+    return lazy.allowedCollectionsFromDump;
+  },
+
   CHANGES_PATH: "/buckets/monitor/collections/changes/changeset",
 
   /**
@@ -450,7 +471,9 @@ export var Utils = {
     if (lastModified === undefined) {
       const { timestamp: dumpTimestamp } = await lazy.SharedUtils.loadJSONDump(
         bucket,
-        collection
+        collection,
+        this.allowedCollections,
+        this.allowedCollectionsFromDump
       );
       // Client recognize -1 as missing dump.
       lastModified = dumpTimestamp ?? -1;
@@ -560,7 +583,9 @@ export var Utils = {
     }
 
     return {
-      changes,
+      changes: changes.filter(change =>
+        lazy.SharedUtils.isCollectionAllowed(change.bucket, change.collection, this.allowedCollections)
+      ),
       timestamp,
       serverTimeMillis,
       backoffSeconds,
diff --git a/services/settings/remote-settings.sys.mjs b/services/settings/remote-settings.sys.mjs
index b180cfc850ee3..41e7f9c43f45d 100644
--- a/services/settings/remote-settings.sys.mjs
+++ b/services/settings/remote-settings.sys.mjs
@@ -19,6 +19,7 @@ ChromeUtils.defineESModuleGetters(lazy, {
   SyncHistory: "resource://services-settings/SyncHistory.sys.mjs",
   UptakeTelemetry: "resource://services-settings/UptakeTelemetry.sys.mjs",
   Utils: "resource://services-settings/Utils.sys.mjs",
+  SharedUtils: "resource://services-settings/SharedUtils.sys.mjs",
 });
 
 const PREF_SETTINGS_BRANCH = "services.settings.";
@@ -286,6 +287,10 @@ function remoteSettingsFunction() {
         const collection = changeset.metadata.id;
         const identifier = `${bucket}/${collection}`;
 
+        if (!lazy.SharedUtils.isCollectionAllowed(bucket, collection, lazy.Utils.allowedCollections)) {
+          continue;
+        }
+
         if (pulled.includes(identifier)) {
           // The startup bundles contain both main and preview changesets.
           // Importing both increases complexity down the line, and brings no value.
diff --git a/toolkit/components/search/ConfigSearchEngine.sys.mjs b/toolkit/components/search/ConfigSearchEngine.sys.mjs
index 5ca99c812c395..f60507b1c084a 100644
--- a/toolkit/components/search/ConfigSearchEngine.sys.mjs
+++ b/toolkit/components/search/ConfigSearchEngine.sys.mjs
@@ -189,7 +189,7 @@ class IconHandler {
       try {
         iconData = await this.#iconCollection.attachments.download(record);
       } catch (ex) {
-        console.error("Could not download new icon", ex);
+        lazy.logConsole.debug("Could not download new icon", ex);
         continue;
       }
 
diff --git a/toolkit/components/search/SearchEngineSelector.sys.mjs b/toolkit/components/search/SearchEngineSelector.sys.mjs
index 3a42b383af198..e375db546bcda 100644
--- a/toolkit/components/search/SearchEngineSelector.sys.mjs
+++ b/toolkit/components/search/SearchEngineSelector.sys.mjs
@@ -409,13 +409,13 @@ export class SearchEngineSelector {
    *   could be obtained.
    */
   async #getConfigurationOverrides() {
-    let result = [];
+/*    let result = [];
     try {
       result = await this.#remoteConfigOverrides.get();
     } catch (ex) {
       // This data is remote only, so we just return an empty array if it fails.
-    }
-    return result;
+    } */
+    return [];
   }
 
   /**
-- 
